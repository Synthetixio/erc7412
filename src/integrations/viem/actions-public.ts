import * as viem from 'viem'

import {
  prepareTransactionRequest as actionPrepareTransactionRequest,
  estimateContractGas as actionEstimateContractGas,
  simulateContract as actionSimulateContract
} from 'viem/actions'

import { callWithOffchainData } from '../..'
import type { OracleAdapter } from '../../types'

/**
 * Extend your viem client with the object returned by this function to automatically apply erc7412
 * required offchain data to your read calls
 */
export function createErc7412PublicActions (adapters: OracleAdapter[]) {
  return (client: viem.PublicClient) => {
    const actions = {
      call: async (args: viem.CallParameters): Promise<viem.CallReturnType> => {
        return {
          data: (
            await callWithOffchainData(
              [
                {
                  from: (args.account as any)?.address ?? viem.zeroAddress,
                  ...args
                }
              ],
              client,
              adapters
            )
          )[0]
        }
      },
      readContract: async (args: viem.ReadContractParameters): Promise<viem.ReadContractReturnType> => {
        return {
          data: viem.decodeFunctionResult({
            ...args,
            data: (
              await callWithOffchainData(
                [
                  {
                    from: (args.account as any)?.address ?? viem.zeroAddress,
                    data: viem.encodeFunctionData(args),
                    ...args
                  }
                ],
                client,
                adapters
              )
            )[0]
          })
        }
      },
      prepareTransactionRequest: async (args: viem.PrepareTransactionRequestParameters) => {
        try {
          return await actionPrepareTransactionRequest(client, args)
        } catch (err) {
          console.log('WARN: erc7412 not implemented for prepareTransactionRequest')
          throw err
        }
      },
      estimateContractGas: async (args: viem.EstimateContractGasParameters) => {
        try {
          return await actionEstimateContractGas(client, args)
        } catch (err) {
          console.log('WARN: erc7412 not implemented for estimateContractGas')
          throw err
        }
      },
      // TODO: types
      simulateContract: async (args: viem.SimulateContractParameters): Promise<any> => {
        try {
          return await actionSimulateContract(client, args)
        } catch (err) {
          const baseTxn = {
            from: (args.account as any)?.address ?? viem.zeroAddress,
            to: args.address,
            chain: args.chain,
            data: viem.encodeFunctionData(args),
            value: args.value
          }
          return {
            result: viem.decodeFunctionResult({
              ...args,
              data: (await callWithOffchainData([baseTxn], client, adapters))[0]
            }) as any
          }
        }
      },
      multicall: async (args: viem.MulticallParameters): Promise<viem.MulticallReturnType> => {
        if (args.contracts.length < 1) {
          throw new Error('must have at least one call for multicall')
        }

        const retvals = await callWithOffchainData(
          // todo: types have a problem with the fact it cannot verify that the array is at least 1 long
          args.contracts.map((c) => {
            return {
              from: c.address ?? viem.zeroAddress,
              data: viem.encodeFunctionData(c),
              ...c
            }
          }) as any,
          client,
          adapters
        )

        return retvals.map((r, i) => {
          return { result: viem.decodeFunctionResult({ ...args.contracts[i], data: r }), status: 'success' }
        })
      }
      // below functions are not included because they are not applicable
      // writeContract: async (args: viem.WriteContractParameters) => {},
    }

    return actions
  }
}
