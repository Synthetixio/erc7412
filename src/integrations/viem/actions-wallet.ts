import * as viem from 'viem'

import { prepareTransactionRequest as actionPrepareTransactionRequest } from 'viem/actions'

import { resolvePrependTransaction, makeTrustedForwarderMulticall, callWithOffchainData } from '../..'

import ITrustedMulticallForwarder from '../../../out/ITrustedMulticallForwarder.sol/ITrustedMulticallForwarder.json'

import type { TransactionRequest } from '../..'
import type { OracleAdapter } from '../../types'
import { getAccount } from './actions-public'

/**
 * Extend your viem client with the object returned by this function to automatically apply erc7412
 * required offchain data to your read calls
 */
export function createErc7412WalletActions (adapters: OracleAdapter[]) {
  return (client: viem.PublicClient) => {
    const actions = {
      prepareTransactionRequest: async (
        args: viem.PrepareTransactionRequestParameters
      ): Promise<viem.PrepareTransactionRequestReturnType> => {
        return await actions.prepareMulticallTransactionRequest({ txns: [args] })
      },

      prepareMulticallTransactionRequest: async (args: {
        txns: viem.PrepareTransactionRequestParameters[]
      }): Promise<viem.PrepareTransactionRequestReturnType> => {
        const prependedTxns: TransactionRequest[] = []

        if (args.txns.length < 1) {
          throw new Error('must have at least 1 transaction in multicall')
        }

        const payloadTxns: TransactionRequest[] = args.txns.map((t) => {
          const req: TransactionRequest = {
            ...t,
            from: getAccount(t.account)
          }

          return req
        })

        while (true) {
          const multicallTxn: viem.PrepareTransactionRequestParameters =
            payloadTxns.length > 1 || prependedTxns.length > 0
              ? {
                  ...makeTrustedForwarderMulticall([...prependedTxns, ...payloadTxns]),
                  account: args.txns[0].account,
                  chain: args.txns[0].chain
                }
              : args.txns[0]
          try {
            return await actionPrepareTransactionRequest(client, multicallTxn)
          } catch (err) {
            prependedTxns.push(await resolvePrependTransaction(err, client, adapters))
          }
        }
      },
      simulateContract: async (args: viem.SimulateContractParameters): Promise<viem.SimulateContractReturnType> => {
        const baseTxn = {
          from: (args.account as any)?.address ?? viem.zeroAddress,
          to: args.address,
          chain: args.chain,
          data: viem.encodeFunctionData(args),
          value: args.value
        }
        const preparedTxn = await actions.prepareTransactionRequest(baseTxn)
        return {
          request:
            preparedTxn.to === args.address
              ? args
              : {
                  ...preparedTxn,
                  abi: ITrustedMulticallForwarder.abi,
                  address: preparedTxn.to,
                  functionName: 'aggregate3Value',
                  args: viem.decodeFunctionData({ abi: ITrustedMulticallForwarder.abi, data: preparedTxn.data ?? '0x' })
                },
          result: viem.decodeFunctionResult({
            ...args,
            data: (await callWithOffchainData([baseTxn], client, adapters))[0]
          })
        } as any // TODO
      }

      // NOTE: cant override `sendRawTransaction` because the transaction is
      // already serialized and so its impossible to impart a new multicall
      // in front without breaking permissions
      // `sendTransaction` is not included because its behavior is not supposed to affect the transaction,
      // its just supposed to sign and send
    }

    return actions
  }
}
