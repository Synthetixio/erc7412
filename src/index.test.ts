import * as mod from './'

import * as viem from 'viem'

import IERC7412 from '../out/IERC7412.sol/IERC7412.json'

import type { OracleAdapter } from './types'

export const fakeWeb3 = {
  request: jest.fn()
}

const fakeAddress = viem.getContractAddress({ from: viem.zeroAddress, nonce: 0n })

export const fakeAdapters: OracleAdapter[] = [
  {
    getOracleId: () => 'FAKE',
    fetchOffchainData: async (_client, _oracleContract, oracleQuery: any[]) =>
      oracleQuery.map((v) => ({ arg: ('0x8765' + v.query.slice(2)) as viem.Hex, fee: BigInt(100) }))
  }
]

describe('index.ts', () => {
  let errorCode: viem.Hex = '0x1234'
  beforeEach(() => {
    errorCode = '0x1234'
    fakeWeb3.request.mockImplementation(async ({ method, params }: { method: string; params: any[] }) => {
      if (method === 'eth_chainId') {
        return BigInt(1337)
      } else if (method === 'eth_call') {
        if (params[0].data === viem.encodeFunctionData({ abi: IERC7412.abi, functionName: 'oracleId' })) {
          // getOracleId call
          return viem.encodeFunctionResult({
            abi: IERC7412.abi,
            functionName: 'oracleId',
            result: viem.stringToHex('FAKE', { size: 32 })
          })
        } else if (params[0].data.includes('87651234') === true) {
          return viem.encodeAbiParameters(
            [viem.parseAbiParameter('(bool success, bytes returnData)[]')],
            [
              [
                { success: true, returnData: '0x1234' },
                { success: true, returnData: '0x5678' }
              ]
            ]
          )
        } else {
          /* eslint @typescript-eslint/no-throw-literal: "off" */
          throw {
            data: viem.encodeErrorResult({
              abi: IERC7412.abi,
              errorName: 'OracleDataRequired',
              args: ['0x2345234523452345234523452345234523452345', errorCode, BigInt(0)]
            })
          }
        }
      } else {
        return '0x'
      }
    })
  })

  describe('makeTrustedForwarderMulticall()', () => {
    it('works even if data and value are unset', async () => {
      expect(mod.makeTrustedForwarderMulticall([{ from: viem.zeroAddress }])).toMatchSnapshot()
    })
  })

  describe('callWithOffchainData()', () => {
    it('passes a call execution error if its not recognized', async () => {
      const origError = new Error('0x08273020')
      fakeWeb3.request.mockRejectedValue(origError)
      await expect(
        async () =>
          await mod.callWithOffchainData(
            [{ from: '0x', to: '0x1234123412341234123412341234123412341234', data: '0x12345678' }],
            fakeWeb3,
            fakeAdapters
          )
      ).rejects.toThrowErrorMatchingSnapshot()
    })

    it('resolves offchain data and returns correct data', async () => {
      expect(
        await mod.callWithOffchainData(
          [
            { from: '0x', to: '0x1234123412341234123412341234123412341234', data: '0x12345678' },
            { from: '0x', to: '0x1234123412341234123412341234123412341234', data: '0x23456789' }
          ],
          fakeWeb3,
          fakeAdapters
        )
      ).toMatchObject([
        { returnData: '0x1234', success: true },
        { returnData: '0x5678', success: true }
      ])
    })

    it('fails if call repeat exceeded', async () => {
      // modify the error code so that the OracleDataRequired error continues forever.
      errorCode = '0x5757'

      await expect(
        mod.callWithOffchainData(
          [
            { from: '0x', to: '0x1234123412341234123412341234123412341234', data: '0x12345678' },
            { from: '0x', to: '0x1234123412341234123412341234123412341234', data: '0x23456789' }
          ],
          fakeWeb3,
          fakeAdapters
        )
      ).rejects.toThrow(new Error('erc7412 callback repeat exceeded'))
    })

    it('fails if no data in call response', async () => {
      fakeWeb3.request.mockResolvedValue(undefined)

      await expect(
        mod.callWithOffchainData(
          [
            { from: '0x', to: '0x1234123412341234123412341234123412341234', data: '0x12345678' },
            { from: '0x', to: '0x1234123412341234123412341234123412341234', data: '0x23456789' }
          ],
          fakeWeb3,
          fakeAdapters
        )
      ).rejects.toThrow(new Error('missing return data from multicall'))
    })
  })

  describe('resolvePrependTransaction()', () => {
    it('passes though error if its not recognized', async () => {
      const origError = '0x08273020'
      await expect(async () => await mod.resolvePrependTransaction(origError, fakeWeb3, fakeAdapters)).rejects.toThrow(
        new Error(`could not parse error. can it be decoded elsewhere? "${origError}"`)
      )
    })

    it('fetches offchain data without fee', async () => {
      const origError = {
        data: viem.encodeErrorResult({
          abi: mod.LEGACY_ODR_ERROR,
          errorName: 'OracleDataRequired',
          args: [fakeAddress, '0x1234']
        })
      }
      fakeWeb3.request.mockResolvedValue(viem.stringToHex('FAKE', { size: 32 }))
      const result = await mod.resolvePrependTransaction(origError, fakeWeb3, fakeAdapters)
      expect(result[0].data).toEqual(
        viem.encodeFunctionData({ abi: IERC7412.abi, functionName: 'fulfillOracleQuery', args: ['0x87651234'] })
      )
    })

    it('fetches offchain data with fee', async () => {
      const origError = {
        error: {
          data: viem.encodeErrorResult({
            abi: IERC7412.abi,
            errorName: 'OracleDataRequired',
            args: [fakeAddress, '0x1234', BigInt(100)]
          })
        }
      }
      fakeWeb3.request.mockResolvedValueOnce(viem.stringToHex('FAKE', { size: 32 }))
      const result = await mod.resolvePrependTransaction(origError, fakeWeb3, fakeAdapters)
      expect(result[0].data).toEqual(
        viem.encodeFunctionData({ abi: IERC7412.abi, functionName: 'fulfillOracleQuery', args: ['0x87651234'] })
      )
      expect(Number(result[0].value)).toEqual(100)
    })

    it('fetches offchain data with multiple errors', async () => {
      const origSubError1 = viem.encodeErrorResult({
        abi: IERC7412.abi,
        errorName: 'OracleDataRequired',
        args: [fakeAddress, '0x1234', BigInt(100)]
      })

      const origSubError2 = viem.encodeErrorResult({
        abi: mod.LEGACY_ODR_ERROR,
        errorName: 'OracleDataRequired',
        args: [fakeAddress, '0x3456']
      })

      const origErrors = {
        data: viem.encodeErrorResult({
          abi: IERC7412.abi,
          errorName: 'Errors',
          args: [[origSubError1, origSubError2]]
        })
      }
      fakeWeb3.request.mockResolvedValue(viem.stringToHex('FAKE', { size: 32 }))
      const result = await mod.resolvePrependTransaction(origErrors, fakeWeb3, fakeAdapters)
      expect(result[0].data).toEqual(
        viem.encodeFunctionData({ abi: IERC7412.abi, functionName: 'fulfillOracleQuery', args: ['0x87651234'] })
      )
      expect(result[1].data).toEqual(
        viem.encodeFunctionData({ abi: IERC7412.abi, functionName: 'fulfillOracleQuery', args: ['0x87653456'] })
      )
      expect(Number(result[0].value)).toEqual(100)
      expect(Number(result[1].value)).toEqual(100)
    })

    it('fails if adapter does not exist', async () => {
      const origError = {
        data: viem.encodeErrorResult({
          abi: IERC7412.abi,
          errorName: 'OracleDataRequired',
          args: [fakeAddress, '0x1234', BigInt(100)]
        })
      }
      fakeWeb3.request.mockResolvedValueOnce(viem.stringToHex('FAKER', { size: 32 }))
      expect(mod.resolvePrependTransaction(origError, fakeWeb3, fakeAdapters)).rejects.toThrow(
        'oracle FAKER not supported (supported oracles: FAKE)'
      )
    })
  })
})
