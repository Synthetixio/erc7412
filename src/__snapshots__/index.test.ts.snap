// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index.ts callWithOffchainData() passes a call execution error if its not recognized 1`] = `"could not parse error. can it be decoded elsewhere? {"cause":{"details":"0x08273020","shortMessage":"An unknown RPC error occurred.","version":"2.21.9","name":"UnknownRpcError","code":-1},"details":"0x08273020","metaMessages":["Raw Call Arguments:","  to:     0xE2C5658cC5C448B48141168f3e475dF8f65A1e3e\\n  value:  0 ETH\\n  data:   0x174dea71000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000020000000000000000000000000123412341234123412341234123412341234123400000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000041234567800000000000000000000000000000000000000000000000000000000"],"shortMessage":"An unknown RPC error occurred.","version":"2.21.9","name":"CallExecutionError"}"`;

exports[`index.ts makeTrustedForwarderMulticall() works even if data and value are unset 1`] = `
{
  "data": "0x174dea7100000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000000",
  "from": "0x0000000000000000000000000000000000000000",
  "to": "0xE2C5658cC5C448B48141168f3e475dF8f65A1e3e",
  "value": 0n,
}
`;
